### PR TITLE
22263 - EFT Refund AP Job Implementation

### PR DESCRIPTION
**Pull request by rodrigo-barraza**
*Issue #:*
https://github.com/bcgov/entity/issues/22263

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 1]
```diff
@@ -172,13 +180,17 @@ def _get_oracle_invoice_batch_name(cls, invoice_number):
...
+        if cls.ap_type == EjvFileType.EFT_REFUND:
+            return f'REFUND_FAS_RS_{invoice_number}'[:30]
         raise RuntimeError('ap_type not selected.')
 
     @classmethod
     def _get_line_code(cls, ap_line: APLine):
         # Routing slip refunds always DEBIT the internal GL and mails out cheques.
         if cls.ap_type == EjvFileType.REFUND:
             return 'D'
+        if cls.ap_type == EjvFileType.EFT_REFUND:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:192]: Use match please, 

I think this is supposed to be 'D' not 'C'

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 2]
```diff
@@ -172,13 +180,17 @@ def _get_oracle_invoice_batch_name(cls, invoice_number):
             return f'{invoice_number}'[:30]
         if cls.ap_type == EjvFileType.REFUND:
             return f'REFUND_FAS_RS_{invoice_number}'[:30]
+        if cls.ap_type == EjvFileType.EFT_REFUND:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:183]: use match please

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 3]
```diff
@@ -163,6 +169,8 @@ def _distribution(cls, distribution_code: str = None):
             return f'{distribution_code}0000000000{cls.EMPTY:<16}'
         if cls.ap_type == EjvFileType.REFUND:
             return f"{current_app.config.get('CGI_AP_DISTRIBUTION')}{cls.EMPTY:<16}"
+        if cls.ap_type == EjvFileType.EFT_REFUND:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:172]: use match please

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 4]
```diff
@@ -144,6 +148,8 @@ def _supplier_location(cls):
             return f"{current_app.config.get('BCA_SUPPLIER_LOCATION'):<3}"
         if cls.ap_type == EjvFileType.REFUND:
             return f"{current_app.config.get('CGI_AP_SUPPLIER_LOCATION'):<3}"
+        if cls.ap_type == EjvFileType.EFT_REFUND:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:151]: use match please

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 5]
```diff
@@ -135,6 +137,8 @@ def _dist_vendor(cls):
             return f"{current_app.config.get('BCA_SUPPLIER_NUMBER'):<30}"
         if cls.ap_type == EjvFileType.REFUND:
             return f"{current_app.config.get('CGI_AP_SUPPLIER_NUMBER'):<30}"
+        if cls.ap_type == EjvFileType.EFT_REFUND:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:140]: use match please

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 6]
```diff
@@ -126,6 +126,8 @@ def _supplier_number(cls):
             return f"{current_app.config.get('BCA_SUPPLIER_NUMBER'):<9}"
         if cls.ap_type == EjvFileType.REFUND:
             return f"{current_app.config.get('CGI_AP_SUPPLIER_NUMBER'):<9}"
+        if cls.ap_type == EjvFileType.EFT_REFUND:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:129]: use match please

File: jobs/payment-jobs/tasks/ap_task.py [hunk 7]
```diff
@@ -59,9 +63,57 @@ def create_ap_files(cls):
...
+            .join(EFTCreditInvoiceLinkModel, EFTCreditInvoiceLinkModel.invoice_id == InvoiceModel.id) \
+            .join(EFTCreditModel, EFTCreditModel.id == EFTCreditInvoiceLinkModel.eft_credit_id) \
+            .join(EFTShortnameLinksModel, EFTShortnameLinksModel.eft_short_name_id == EFTCreditModel.short_name_id) \
+            .join(EFTRefundModel, EFTRefundModel.short_name_id == EFTShortnameLinksModel.eft_short_name_id) \
+            .filter(EFTRefundModel.status == InvoiceStatus.REFUND_REQUESTED.value) \
+            .filter(EFTRefundModel.refund_amount > 0) \
+            .all()
+
+        current_app.logger.info(f'Found {len(eft_refunds_dao)} to refund.')
+        if not eft_refunds_dao:
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:89]: 89, 90 probably not needed

File: jobs/payment-jobs/tasks/ap_task.py [hunk 8]
```diff
@@ -59,9 +63,57 @@ def create_ap_files(cls):
...
+            return
+
+        for refunds in list(batched(eft_refunds_dao, 250)):
+            ejv_file_model: EjvFileModel = EjvFileModel(
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:93]: no type hint needed here, compiler will infer it

File: jobs/payment-jobs/tasks/ap_task.py [hunk 9]
```diff
@@ -59,9 +63,57 @@ def create_ap_files(cls):
...
+            for invoice_refund in refunds:
+                invoice, eft_refund = invoice_refund
+                current_app.logger.info(f'Creating refund for {invoice.id}, Amount {eft_refund.refund_amount}.')
+                ap_content = f'{ap_content}{cls.get_ap_header(
+                    eft_refund.refund_amount, invoice.id, invoice.created_on)}'
+                ap_line = APLine(total=eft_refund.refund_amount, invoice_number=invoice.id, line_number=1)
+                ap_content = f'{ap_content}{cls.get_ap_invoice_line(ap_line)}'
+                line_count_total += 2
+                batch_total += eft_refund.refund_amount
+                eft_refund.status = InvoiceStatus.REFUNDED.value
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:112]: It's not really refunded until we receive the feedback with 0000

File: jobs/payment-jobs/tasks/ap_task.py (hunk 7 above)
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:89]: 89, 90 not needed

File: jobs/payment-jobs/tasks/ap_task.py [hunk 10]
```diff
@@ -59,9 +63,57 @@ def create_ap_files(cls):
...
+            batch_number: str = cls.get_batch_number(ejv_file_model.id)
+            ap_content: str = cls.get_batch_header(batch_number)
+            batch_total = 0
+            line_count_total: int = 0
+            for invoice_refund in refunds:
+                invoice, eft_refund = invoice_refund
+                current_app.logger.info(f'Creating refund for {invoice.id}, Amount {eft_refund.refund_amount}.')
+                ap_content = f'{ap_content}{cls.get_ap_header(
+                    eft_refund.refund_amount, invoice.id, invoice.created_on)}'
+                ap_line = APLine(total=eft_refund.refund_amount, invoice_number=invoice.id, line_number=1)
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:108]: line_number=1?

File: jobs/payment-jobs/tasks/ap_task.py [hunk 11]
```diff
@@ -59,9 +63,58 @@ def create_ap_files(cls):
...
+    @classmethod
+    def _create_eft_refund_file(cls):
+        """Create AP file for EFT refunds and upload to CGI."""
+        cls.ap_type = EjvFileType.EFT_REFUND
+        eft_refunds_dao: List[tuple[InvoiceModel, EFTRefundModel]] = db.session.query(InvoiceModel, EFTRefundModel) \
+            .join(EFTCreditInvoiceLinkModel, EFTCreditInvoiceLinkModel.invoice_id == InvoiceModel.id) \
+            .join(EFTCreditModel, EFTCreditModel.id == EFTCreditInvoiceLinkModel.eft_credit_id) \
+            .join(EFTShortnameLinksModel, EFTShortnameLinksModel.eft_short_name_id == EFTCreditModel.short_name_id) \
+            .join(EFTRefundModel, EFTRefundModel.short_name_id == EFTShortnameLinksModel.eft_short_name_id) \
+            .filter(EFTRefundModel.status == InvoiceStatus.REFUND_REQUESTED.value) \
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:84]: Need to filter out eft_refund.disbursement_status_code = UPLOADED ?

File: jobs/payment-jobs/tasks/ap_task.py [hunk 12]
```diff
@@ -59,9 +63,58 @@ def create_ap_files(cls):
...
+        current_app.logger.info(f'Found {len(eft_refunds_dao)} to refund.')
+
+        for refunds in list(batched(eft_refunds_dao, 250)):
+            ejv_file_model = EjvFileModel(
+                file_type=cls.ap_type.value,
+                file_ref=cls.get_file_name(),
+                disbursement_status_code=DisbursementStatus.UPLOADED.value
+            ).flush()
+
+            batch_number: str = cls.get_batch_number(ejv_file_model.id)
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:97]: change hint for get_batch_number to return `-> str`

File: jobs/payment-jobs/tasks/ap_task.py [hunk 13]
```diff
@@ -59,9 +63,58 @@ def create_ap_files(cls):
...
+            ap_content: str = cls.get_batch_header(batch_number)
+            batch_total = 0
+            line_count_total: int = 0
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py:100]: prob don't need int, = 0 already indicates? I think you could do 0.0 for float, Decimal('0') for decimal

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 14]
```diff
@@ -122,29 +122,35 @@ def get_ap_comment(cls, refund_details, routing_slip_number):
...
-        if cls.ap_type == EjvFileType.NON_GOV_DISBURSEMENT:
-            return f"{current_app.config.get('BCA_SUPPLIER_NUMBER'):<30}"
-        if cls.ap_type == EjvFileType.REFUND:
-            return f"{current_app.config.get('CGI_AP_SUPPLIER_NUMBER'):<30}"
-        raise RuntimeError('ap_type not selected.')
+        match cls.ap_type:
+            case EjvFileType.NON_GOV_DISBURSEMENT:
+                return f"{current_app.config.get('BCA_SUPPLIER_NUMBER'):<30}"
+            case EjvFileType.REFUND | EjvFileType.EFT_REFUND:
+                return f"{current_app.config.get('CGI_AP_SUPPLIER_NUMBER'):<30}"
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:140]: These will be different GL's, you'll need a new config line for it I think
`EFT_AP_SUPPLIER_NUMBER`

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 15]
```diff
@@ -122,29 +122,35 @@ def get_ap_comment(cls, refund_details, routing_slip_number):
...
-        if cls.ap_type == EjvFileType.NON_GOV_DISBURSEMENT:
-            return f"{current_app.config.get('BCA_SUPPLIER_LOCATION'):<3}"
-        if cls.ap_type == EjvFileType.REFUND:
-            return f"{current_app.config.get('CGI_AP_SUPPLIER_LOCATION'):<3}"
-        raise RuntimeError('ap_type not selected.')
+        match cls.ap_type:
+            case EjvFileType.NON_GOV_DISBURSEMENT:
+                return f"{current_app.config.get('BCA_SUPPLIER_LOCATION'):<3}"
+            case EjvFileType.REFUND | EjvFileType.EFT_REFUND:
+                return f"{current_app.config.get('CGI_AP_SUPPLIER_LOCATION'):<3}"
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:151]: same as above, this needs to be new I think

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 16]
```diff
@@ -159,28 +165,32 @@ def _get_date(cls, date):
...
-        if cls.ap_type == EjvFileType.NON_GOV_DISBURSEMENT:
-            return f'{distribution_code}0000000000{cls.EMPTY:<16}'
-        if cls.ap_type == EjvFileType.REFUND:
-            return f"{current_app.config.get('CGI_AP_DISTRIBUTION')}{cls.EMPTY:<16}"
-        raise RuntimeError('ap_type not selected.')
+        match cls.ap_type:
+            case EjvFileType.NON_GOV_DISBURSEMENT:
+                return f'{distribution_code}0000000000{cls.EMPTY:<16}'
+            case EjvFileType.REFUND | EjvFileType.EFT_REFUND:
+                return f"{current_app.config.get('CGI_AP_DISTRIBUTION')}{cls.EMPTY:<16}"
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:172]: Needs to be new

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 17]
```diff
@@ -159,28 +165,32 @@ def _get_date(cls, date):
...
-        if cls.ap_type == EjvFileType.NON_GOV_DISBURSEMENT:
-            return f'{invoice_number}'[:30]
-        if cls.ap_type == EjvFileType.REFUND:
-            return f'REFUND_FAS_RS_{invoice_number}'[:30]
-        raise RuntimeError('ap_type not selected.')
+        match cls.ap_type:
+            case EjvFileType.NON_GOV_DISBURSEMENT:
+                return f'{invoice_number}'[:30]
+            case EjvFileType.REFUND | EjvFileType.EFT_REFUND:
+                return f'REFUND_FAS_RS_{invoice_number}'[:30]
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:183]: EFT_REFUND has nothing to do with FAS

File: jobs/payment-jobs/tasks/common/cgi_ap.py [hunk 18]
```diff
@@ -122,29 +122,35 @@ def get_ap_comment(cls, refund_details, routing_slip_number):
...
-        if cls.ap_type == EjvFileType.NON_GOV_DISBURSEMENT:
-            return f"{current_app.config.get('BCA_SUPPLIER_NUMBER'):<9}"
-        if cls.ap_type == EjvFileType.REFUND:
-            return f"{current_app.config.get('CGI_AP_SUPPLIER_NUMBER'):<9}"
-        raise RuntimeError('ap_type not selected.')
+        match cls.ap_type:
+            case EjvFileType.NON_GOV_DISBURSEMENT:
+                return f"{current_app.config.get('BCA_SUPPLIER_NUMBER'):<9}"
+            case EjvFileType.REFUND | EjvFileType.EFT_REFUND:
+                return f"{current_app.config.get('CGI_AP_SUPPLIER_NUMBER'):<9}"
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/cgi_ap.py:129]: Probably needs to be new, you'll need to get this info from Finance.

I think it's probably the GL for EFT holding account, but would need to confirm with Finance

File: jobs/payment-jobs/tests/jobs/test_ap_task.py [hunk 19]
```diff
@@ -23,11 +23,40 @@
...
+    eft_transaction_id = factory_create_eft_transaction(file_id=eft_file.id).id
+    short_name_id = factory_create_eft_shortname('shortnametest').id
+    factory_create_eft_refund(short_name_id=short_name_id)
+    eft_credit = factory_create_eft_credit(short_name_id=short_name_id, eft_transaction_id=eft_transaction_id)
+    factory_create_eft_credit_invoice_link(invoice_id=invoice.id, eft_credit_id=eft_credit.id)
+    factory_eft_shortname_link(short_name_id=short_name_id)
+
+    with patch('pysftp.Connection.put') as mock_upload:
+        ApTask.create_ap_files()
+        mock_upload.assert_called()
```
Comment (seeker25) [inline jobs/payment-jobs/tests/jobs/test_ap_task.py:57]: Going to need some unit tests where you look at specific positioning for EFT

Comment (seeker25) [reply jobs/payment-jobs/tests/jobs/test_ap_task.py:57]: Hmm I see the existing tests aren't anything like the EJV unit tests where it actually tests the files positions etc

Comment (seeker25) [reply jobs/payment-jobs/tests/jobs/test_ap_task.py:57]: So you'll have to run a few batches on TEST and look at the feedback instead at the very minimum

File: pay-api/migrations/versions/2024_09_04_6ce240c993e6_.py [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+# Please ignore those lines and don't include in migration.
+
+revision = '6ce240c993e6'
+down_revision = '2097573390f1'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.execute("set statement_timeout=900000;")
```
Comment (seeker25) [inline pay-api/migrations/versions/2024_09_04_6ce240c993e6_.py:24]: prob don't need this, eft_refunds traffic is non existent

**Comment by seeker25** [comment]
You'll need to:

1. Grab an existing AP file and figure out the bits that are specific to routing slip refunds that need to be changed into EFT refunds, confirm with finance the fields that should go in there

2. After you've done the above, grab the secrets from TEST and submit a file on TEST and make sure it returns back 0000

Note: some of the GL's that are on PROD might not work on TEST, so look at the existing secrets in TEST for AP and compare

File: jobs/payment-jobs/tasks/ap_task.py (hunk 11 above)
Comment (ochiu) [reply jobs/payment-jobs/tasks/ap_task.py:84]: EFTRefundModels.status probably shouldn't be compared to the InvoiceStatus.

It also looks like when we create the eft_refund in pay api it uses EFTCreditInvoiceStatus. EFT refunds has its own flow, I would suggest it has its own set of statuses.

Something similar to:

PENDING_APPROVAL
REJECTED
APPROVED

**Review by seeker25 — APPROVED** [review]

File: pay-api/src/pay_api/models/eft_refund.py [hunk 21]
```diff
@@ -58,6 +59,8 @@ class EFTRefund(Audit):
     comment = db.Column(db.String(), nullable=False)
     decline_reason = db.Column(db.String(), nullable=True)
     created_by = db.Column('created_by', db.String(100), nullable=True)
+    created_on = db.Column('created_on', db.DateTime, nullable=False, default=lambda: datetime.now(tz=timezone.utc))
+    disbursement_status_code = db.Column(db.String(20), ForeignKey('disbursement_status_codes.code'), nullable=True)
```
Comment (seeker25) [inline pay-api/src/pay_api/models/eft_refund.py:63]: Looks good I'll clean this one out later

**Pull request merged**